compiler/src/dotty/tools/dotc/core/CheckRealizable.scala

```diff
@@ -72,16 +72,18 @@ class CheckRealizable(implicit ctx: Context) {
   def realizability(tp: Type): Realizability = tp.dealias match {
     case tp: TermRef =>
       val sym = tp.symbol
-      if (sym.is(Stable)) realizability(tp.prefix)
-      else {
-        val r =
-          if (!sym.isStable) NotStable
-          else if (!isLateInitialized(sym)) realizability(tp.prefix)
-          else if (!sym.isEffectivelyFinal) new NotFinal(sym)
-          else realizability(tp.info).mapError(r => new ProblemInUnderlying(tp.info, r))
-        if (r == Realizable) sym.setFlag(Stable)
-        r
-      }
+      val r =
+        if (sym.is(Stable)) realizability(tp.prefix)
+        else {
+          val r =
+            if (!sym.isStable) NotStable
+            else if (!isLateInitialized(sym)) realizability(tp.prefix)
+            else if (!sym.isEffectivelyFinal) new NotFinal(sym)
+            else realizability(tp.info).mapError(r => new ProblemInUnderlying(tp.info, r))
+          if (r == Realizable) sym.setFlag(Stable)
+          r
+        }
+      if (r == Realizable || tp.info.isStableRealizable) Realizable else r
```

Comment:
`r == Realizable || ` can be dropped without changing meaning. In fact, it's unclear why we should check `tp.info.isStableRealizable` at this point — if this is necessary, it seems to warrant a comment. Maybe this subtly depends on `setFlag` influencing `isStableRealizable`.

Should we check `tp.info.isStableRealizable` first and skip all the other expensive computations? That doesn't set the `sym.is(Stable)` cache, but it can be set on request the next time.*

Otherwise, should `tp.info.isStableRealizable` be used when computing `r`?

Reply:
Playing with this myself.